orchagent/pfcwdorch.cpp

```diff
@@ -446,173 +392,109 @@ bool PfcWdSwOrch<DropHandler, ForwardHandler>::startWdOnQueue(sai_object_id_t qu
 {
     SWSS_LOG_ENTER();
 
-    {
-        unique_lock<mutex> lk(m_pfcWdMutex);
-
-        // Remove from internal DB
-        m_entryMap.erase(queueId);
-    }
-
-    if (m_entryMap.empty())
+    if (m_runPfcWdSwOrchThread.load())
     {
         endWatchdogThread();
     }
 
-    return true;
-}
-
-template <typename DropHandler, typename ForwardHandler>
-uint32_t PfcWdSwOrch<DropHandler, ForwardHandler>::getNearestPollTime(void)
-{
-    SWSS_LOG_ENTER();
-
-    uint32_t nearestTime = 0;
-
-    for (const auto& queueKv : m_entryMap)
-    {
-        const PfcWdQueueEntry& queueEntry = queueKv.second;
-
-        // First check regular polling intervals
-        if (nearestTime == 0 || queueEntry.pollTimeLeft < nearestTime)
-        {
-            nearestTime = queueEntry.pollTimeLeft;
-        }
-    }
+    unregisterFromWdDb(port);
 
-    return nearestTime;
+    return true;
 }
 
 template <typename DropHandler, typename ForwardHandler>
-void PfcWdSwOrch<DropHandler, ForwardHandler>::pollQueues(uint32_t nearestTime, DBConnector& db,
-        string detectSha, string restoreSha)
+void PfcWdSwOrch<DropHandler, ForwardHandler>::handleWdNotification(swss::NotificationConsumer& wdNotification)
 {
     SWSS_LOG_ENTER();
 
-    unique_lock<mutex> lk(m_pfcWdMutex);
-
-    // Select those queues for which timer expired
-    vector<string> normalQueues;
-    vector<string> stormedQueues;
-    for (const auto& queueKv : m_entryMap)
-    {
-        sai_object_id_t queueId = queueKv.first;
-        const PfcWdQueueEntry& queueEntry = queueKv.second;
-
-        if (queueEntry.pollTimeLeft == nearestTime)
-        {
-            // Queue is being stormed
-            if (queueEntry.handler != nullptr)
-            {
-                stormedQueues.push_back(sai_serialize_object_id(queueId));
-            }
-            // Queue is not stormed
-            else
-            {
-                normalQueues.push_back(sai_serialize_object_id(queueId));
-            }
-        }
-    }
+    string queueIdStr;
+    string event;
+    vector<swss::FieldValueTuple> values;
 
-    // Run scripts for selected queues to see if their state changed
-    // from normal to stormed and vice versa
-    set<string> stormCheckReply;
-    set<string> restoreCheckReply;
-    vector<string> argv =
-    {
-        to_string(COUNTERS_DB),
-        COUNTERS_TABLE,
-        to_string(nearestTime * 1000)
-    };
+    wdNotification.pop(queueIdStr, event, values);
 
-    if (!normalQueues.empty())
-    {
-        stormCheckReply = runRedisScript(db, detectSha, normalQueues, argv);
-    }
+    sai_object_id_t queueId = SAI_NULL_OBJECT_ID;
+    sai_deserialize_object_id(queueIdStr, queueId);
 
-    if (!stormedQueues.empty())
+    auto entry = m_entryMap.find(queueId);
+    if (entry == m_entryMap.end())
     {
-        restoreCheckReply = runRedisScript(db, restoreSha, stormedQueues, argv);
+        SWSS_LOG_ERROR("Queue %s is not registered", queueIdStr.c_str());
```

Comment:
should return here since entry is null now.

Reply:
Done